lib/algos/cache/GapSet.ts

```diff
@@ -145,20 +145,19 @@ export default class GapSet implements Iterable<GapSetEntry> {
 
         const lookup = this._lookupOrCreateGap(firstKey, lastKey, weight);
         let { gap: curGap, weightToMerge } = lookup;
-        let findNextGap;
-        if (curGap.firstKey === curGap.lastKey) {
-            // single-key gaps are never chained: find the first gap strictly after 'curGap'
-            // Note: using the 'OrEqual' method is not possible as it would yield 'curGap'
-            findNextGap = this._gaps.findLeastGreaterThan.bind(this._gaps);
-        } else {
-            // gaps with more than one key may be chained (g1.lastKey == g2.firstKey) and we
-            // need the 'OrEqual' method to find those chained gaps
-            findNextGap = this._gaps.findLeastGreaterThanOrEqual.bind(this._gaps);
-        }
         let mergedWeightSum = 0;
         // loop over existing mergeable gaps straddled by 'gap.lastKey' -> 'lastKey'
         while (lastKey > curGap.lastKey) {
-            const nextGap = findNextGap({ firstKey: curGap.lastKey })?.value;
+            let nextGap;
+            if (curGap.firstKey === curGap.lastKey) {
```

Comment:
Gaps do actually overlap (by a single key), that's why we need the `OrEqual` version to find them. But we also need to skip the single-key gap that we may start with hence the two different calls.

Reply:
Sorry I meant strictly overlapping, as in intersecting on more than one key. We agree that the data store should never be in a state where we have strictly overlapping gaps?

While we're at it, can you confirm that we cannot have two gaps with the same `firstKey`? I inferred from the fact that it's a set but I just want confirmation. 

If that's the case, I think my logic works, there are four cases:
I'll write your approach as A#1 and my proposal as A#2:

```
If gap is a single key, and nextGap is chained:
ant..ant - ant..fox
I think this is not possible. 
---
If gap is a single key, and nextGap is not chained:
ant..ant - cat..fox
findLeastGreaterThan('ant') will work (both A#1 and A#2)
---
If gap spans multiple keys, and nextGap is chained:
ant..cat - cat..fox
A#1 - findLeastGreaterThanOrEqual('cat') will work
A#2 - findLeastGreaterThan('ant') will work
---
If gap spans multiple keys, and nextGap is not chained:
ant..cat - elk..fox
A#1 - findLeastGreaterThanOrEqual('cat') will work
A#2 - findLeastGreaterThan('ant') will work
```

Reply:
Ah I see what you mean, comparing against the first key rather than the last key of the current gap. Should work indeed!

Yes gaps can only overlap by a single key in any case, and there can be only one gap starting with a given key.

Reply:
updated